### PR TITLE
[SofaMiscForceField][SofaBaseMechanics] Add compile-time conditions to avoid compiling unrealistic cases

### PR DESCRIPTION
**Pull request by fredroy**
Link with issue #2457

~~The solution here is quite straight-forward (and a bit ugly IMO): encase 2d & 3d cases with `if constexpr (dimensions == X)`
Not ideal for the moment but hard to come by with the current design.~~ now the  methods are not compiled if the dimension is wrong,
The merge of those two masses should this into account. (mentioned here #2452)

Plus remove "static_warning" if the call was wrong and replace it with a real static_assert

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: SofaKernel/modules/Sofa.Geometry/src/sofa/geometry/Tetrahedron.h [hunk 1]
```diff
@@ -54,23 +53,14 @@ struct Tetrahedron
     static constexpr auto volume(const Node& n0, const Node& n1, const Node& n2, const Node& n3)
     {
         constexpr Node n{};
-        //static_assert(std::distance(std::begin(n), std::end(n)) == 3, "volume can only be computed in 3 dimensions.");
+        static_assert(std::distance(std::begin(n), std::end(n)) == 3, "volume can only be computed in 3 dimensions.");
```
Comment (damienmarchal) [inline SofaKernel/modules/Sofa.Geometry/src/sofa/geometry/Tetrahedron.h:56]: should'nt it be a enable_if here ?

Comment (fredroy) [reply SofaKernel/modules/Sofa.Geometry/src/sofa/geometry/Tetrahedron.h:56]: IMO the static_assert version will give a more meaningful message when throwing a compilation error (in my experience, std::enable_if is a bit... confusing when the compilation fails)

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

**Pull request merged**